### ISSUE TITLE
Duplicate violations from WhitespaceAfter and WhitespaceAround on reserved words

### ISSUE THREAD

**Issue by Tieshiyang** [checkstyle/checkstyle#17718]
I have read check documentation: https://checkstyle.sourceforge.io/checks/whitespace/whitespaceafter.html
 and https://checkstyle.sourceforge.io/checks/whitespace/whitespacearound.html
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

From: https://google.github.io/styleguide/javaguide.html#s4.6.2-horizontal-whitespace
> Separating any reserved word, such as if, for or catch, from an open parenthesis (() that follows it on that line.
Separating any reserved word, such as else or catch, from a closing curly brace (}) that precedes it on that line.

Code: 
```java
class Test {
  void testIf(int x) {
    if(x > 0) { // Expected 1 Violation but get 2
      System.out.println("Positive");
    }
  }

  void testFor() {
    for(int i = 0; i < 5; i++) { // Expected 1 Violation but get 2
      System.out.println(i);
    }
  }
}
```

Cli: 
```cmd
$ java -jar checkstyle-10.26.1-all.jar -c google_checks.xml Test.java 
Starting audit...
[WARN] Test.java:3:5: 'if' is not followed by whitespace. [WhitespaceAfter]
[WARN] Test.java:3:5: WhitespaceAround: 'if' is not followed by whitespace. Empty blocks                may only be represented as {} when not part of a multi-block statement (4.1.3) [WhitespaceAround]
[WARN] Test.java:9:5: 'for' is not followed by whitespace. [WhitespaceAfter]
[WARN] Test.java:9:5: WhitespaceAround: 'for' is not followed by whitespace. Empty blocks                may only be represented as {} when not part of a multi-block statement (4.1.3) [WhitespaceAround]
Audit done.
```

Formatter:
```cmd
$ java -jar google-java-format-1.28.0-all-deps.jar Test.java > FormattedCode.java
$ diff -Naru Test.java FormattedCode.java
--- Test.java   2025-09-01 21:15:02.351271000 +0800
+++ FormattedCode.java  2025-09-01 21:23:07.905927300 +0800
@@ -1,12 +1,12 @@
 class Test {
   void testIf(int x) {
-    if(x > 0) { // Expected 1 Violation but get 2
+    if (x > 0) { // Expected 1 Violation but get 2
       System.out.println("Positive");
     }
   }

   void testFor() {
-    for(int i = 0; i < 5; i++) { // Expected 1 Violation but get 2
+    for (int i = 0; i < 5; i++) { // Expected 1 Violation but get 2
       System.out.println(i);
     }
   }
```

---

For `if(x` and `for(int`, both `WhitespaceAfter` and `WhitespaceAround` report the same missing space.
This causes duplicate warnings (2 per violation) though only 1 is expected.

I guess in `google_checks.xml` the intention of using `WhitespaceAfter` is to check the space after reserved words, while `WhitespaceAround` is used to check the space before them.

So perhaps we should provide a clearer separation of responsibilities (e.g. `WhitespaceBefore` / `WhitespaceAfter` only, without duplication).
This would reduce duplicate warnings and false positives in the future.

---

**Comment by romani** [comment]
Kind of the same as https://github.com/checkstyle/checkstyle/issues/17719#issuecomment-3331402412

We consider it as ok that user receives multiple violations on same root problem. It happens sometimes for others Checks too.

Each of them is valid violation.
Not ideal, yes, but it will be not easy to fix it and might result in blind spots. So better to keep potential duplicates, to avoid missed violations.

Issue can not be approved for now.
Please provide more reasons, or we can close issue.

**Comment by VeniceWish** [comment]
Thanks for the explanation.

I was thinking that maybe we could follow the same design pattern as `WhitespaceAfter` to implement a `WhitespaceBefore` check. Then, when configuring `WhitespaceAround`, users could choose to combine them instead, which would avoid duplication and allow each check to operate at a consistent granularity. But that’s just a simple idea — I understand the real implementation might be more complex.

That said, since it is considered acceptable in Checkstyle for users to receive multiple violations for the same root issue, I’m fine with closing this issue.

Thanks again for the thoughtful discussion!
